internal/cmd/list-alerts.go

```diff
@@ -0,0 +1,230 @@
+					bunt.Printf("   *Description:* \n")
+					bunt.Printf("   %s\n", incident.Description)
+				}
+
+				bunt.Printf("   *Link:* %s\n", strings.Replace(incident.Self, "api", "ibm", 1))
+
+				start := mustParsePagerDutyRFC3339ishTime(incident.CreatedAt)
+
+				end := mustParsePagerDutyRFC3339ishTime(incident.LastStatusChangeAt)
+
+				bunt.Printf("   *Time:* %s - %s (%s)\n",
+					start.Local().Format("2006-01-02 15:04:05"),
+					end.Local().Format("2006-01-02 15:04:05"),
+					end.Sub(start),
+				)
+
+				notes, err := client.ListIncidentNotes(incident.Id)
+				if err != nil {
+					return err
+				}
+				if len(notes) > 0 {
+					bunt.Printf("   *Notes:*\n")
+					for j := len(notes) - 1; j >= 0; j-- {
+						bunt.Printf("      %d. %s (by _%s_ at _%s_)\n",
+							len(notes)-j,
+							notes[j].Content,
+							lookUpNameByUserID(client, notes[j].User.ID),
+							formatNoteTime(notes[j].CreatedAt),
+						)
+					}
+				}
+
+				numPrintedIncidents++
+			}
+			if err != nil {
+				return err
+			}
+
+			if a.More {
+				offset += limit
+			} else {
+				moreIncidents = false
+			}
+		}
+
+		bunt.Println()
+
+		return nil
+	},
+}
+
+func lookUpNameByUserID(client *pagerduty.Client, id string) string {
+	user, err := client.GetUser(id, pagerduty.GetUserOptions{})
+	if err != nil {
+		return "unknown"
+	}
+	return user.Name
+}
+
+func formatNoteTime(input string) string {
+	time, err := time.Parse("2006-01-02T15:04:05-07:00", input)
+	if err != nil {
+		return input
+	}
+
+	return time.Local().Format("2006-01-02 15:04:05")
+}
+
+func filterIncidentsByNameInLogEntries(incidents []pagerduty.Incident, username string, client *pagerduty.Client) ([]pagerduty.Incident, error) {
+	numIncidents := len(incidents)
+	var filteredIncidents []pagerduty.Incident
+	logEntries := make([]pagerduty.ListIncidentLogEntriesResponse, numIncidents, numIncidents)
+
+	loadLogEntries := func(position int, wg *sync.WaitGroup) {
+		logEntry, _ := client.ListIncidentLogEntries(incidents[position].Id, pagerduty.ListIncidentLogEntriesOptions{})
+		logEntries[position] = *logEntry
+		wg.Done()
+	}
+	var wg *sync.WaitGroup
+	if numIncidents <= 50 { // if you sent 100 requests at once, you sometimes get an error. 50 requests at once is pretty safe.
+		wg = new(sync.WaitGroup)
+		wg.Add(numIncidents)
+
+		for i := 0; i < numIncidents; i++ {
+			go loadLogEntries(i, wg)
+		}
+	} else {
+		wg = new(sync.WaitGroup)
+		wg.Add(int(numIncidents / 2))
+
+		for i := 0; i < int(numIncidents/2); i++ {
+			go loadLogEntries(i, wg)
+		}
+
+		wg.Wait()
+		wg.Add(int(numIncidents / 2))
+
+		for i := int(numIncidents / 2); i < numIncidents; i++ {
+			go loadLogEntries(i, wg)
+		}
+	}
+	wg.Wait()
+
+	for i, incident := range incidents {
+		for _, logEntry := range logEntries[i].LogEntries {
+			if strings.Contains(logEntry.CommonLogEntryField.Summary, username) {
+				filteredIncidents = append(filteredIncidents, incident)
+				break
+			}
+		}
+	}
+	return filteredIncidents, nil
+}
```

Comment:
Very good that you figured out that you have to rate limit the requests when doing things in parallel. There is a code pattern that solves the same thing with a bit less code and `if` blocks. Have a look at my attached suggestion. What it does is to spin up _n_  worker Go routines that then wait for work tasks to be done. Using a Go channel you can supply work tasks to the workers. A worker finishes when the work task channel is closed. That way you can relatively elegant solve the problem to not overload the back-end.

```suggestion
func filterIncidentsByNameInLogEntries(incidents []pagerduty.Incident, username string, client *pagerduty.Client) ([]pagerduty.Incident, error) {
	const parallel = 10

	type in struct {
		index    int
		incident pagerduty.Incident
	}

	var (
		errors     = []error{}
		tasks      = make(chan in, len(incidents))
		logEntries = make([]pagerduty.ListIncidentLogEntriesResponse, len(incidents))
	)

	// Fill the task channel with work to be done
	for idx, incident := range incidents {
		tasks <- in{idx, incident}
	}

	// Start a worker group to process work tasks
	var wg sync.WaitGroup
	wg.Add(parallel)
	for i := 0; i < parallel; i++ {
		go func() {
			defer wg.Done()
			for task := range tasks {
				resp, err := client.ListIncidentLogEntries(task.incident.Id, pagerduty.ListIncidentLogEntriesOptions{})
				if err != nil {
					errors = append(errors, err)
					break
				}

				logEntries[task.index] = *resp
			}
		}()
	}

	close(tasks)
	wg.Wait()

	if len(errors) > 0 {
		return nil, wrap.Errors(errors, "failed to filter incidents by name")
	}

	var filteredIncidents []pagerduty.Incident
	for i, incident := range incidents {
		for _, logEntry := range logEntries[i].LogEntries {
			if strings.Contains(logEntry.CommonLogEntryField.Summary, username) {
				filteredIncidents = append(filteredIncidents, incident)
				break
			}
		}
	}

	return filteredIncidents, nil
}
```